File: Jiten.Api/Helpers/MediaStatusFilter.cs
using Jiten.Core.Data;

namespace Jiten.Api.Helpers;

/// <summary>The media list's comma-separated status filter; ticked tokens are OR-ed together.</summary>
public sealed class MediaStatusFilter
{
    private static readonly Dictionary<string, DeckStatus> StatusTokens = new()
    {
        ["planning"] = DeckStatus.Planning,
        ["ongoing"] = DeckStatus.Ongoing,
        ["completed"] = DeckStatus.Completed,
        ["dropped"] = DeckStatus.Dropped,
    };

    public HashSet<DeckStatus> Statuses { get; } = [];
    public bool NoStatus { get; private set; }
    public bool Ignored { get; private set; }

    /// <summary>Legacy "fav" token, which predates the standalone favourite flag and must keep working for old URLs and presets.</summary>
    public bool Favourite { get; private set; }

    public bool HasStatusCriteria => Statuses.Count > 0 || NoStatus;

    public static MediaStatusFilter Parse(string? raw)
    {
        var filter = new MediaStatusFilter();
        if (string.IsNullOrWhiteSpace(raw))
            return filter;

        foreach (var token in raw.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            var normalized = token.ToLowerInvariant();
            if (StatusTokens.TryGetValue(normalized, out var status))
                filter.Statuses.Add(status);
            else if (normalized == "nostatus")
                filter.NoStatus = true;
            else if (normalized == "ignore")
                filter.Ignored = true;
            else if (normalized == "fav")
                filter.Favourite = true;
        }

        return filter;
    }

    /// <summary>Keep decks in DeckIds, or drop them when Exclude; "nostatus" must stay an exclusion to cover decks with no preference row.</summary>
    public (HashSet<int> DeckIds, bool Exclude) ResolveDeckIds(IEnumerable<UserDeckPreference> prefs, HashSet<int> ignoredDeckIds)
    {
        if (!HasStatusCriteria && !Ignored)
            return (ignoredDeckIds, true);

        if (NoStatus)
        {
            var excluded = prefs.Where(p => p.Status != DeckStatus.None && !Statuses.Contains(p.Status))
                                .Select(p => p.DeckId)
                                .ToHashSet();
            if (Ignored)
                excluded.ExceptWith(ignoredDeckIds);
            else
                excluded.UnionWith(ignoredDeckIds);
            return (excluded, true);
        }

        var included = prefs.Where(p => Statuses.Contains(p.Status))
                            .Select(p => p.DeckId)
                            .ToHashSet();
        if (Ignored)
            included.UnionWith(ignoredDeckIds);
        else
            included.ExceptWith(ignoredDeckIds);
        return (included, false);
    }
}

File: Jiten.Web/app/utils/mediaStatusFilter.ts
export const MEDIA_STATUS_OPTIONS = [
    {label: 'Planning', value: 'planning'},
    {label: 'Ongoing', value: 'ongoing'},
    {label: 'Completed', value: 'completed'},
    {label: 'Dropped', value: 'dropped'},
    {label: 'Without status', value: 'nostatus'},
    {label: 'Ignored', value: 'ignore'},
] as const;

export type MediaStatusToken = (typeof MEDIA_STATUS_OPTIONS)[number]['value'];

const TOKEN_ORDER: readonly string[] = MEDIA_STATUS_OPTIONS.map((option) => option.value);

export const mediaStatusLabel = (token: string): string => MEDIA_STATUS_OPTIONS.find((option) => option.value === token)?.label ?? 'Status';

export const normaliseStatusTokens = (tokens: readonly string[]): MediaStatusToken[] =>
    TOKEN_ORDER.filter((token) => tokens.includes(token)) as MediaStatusToken[];

export const parseStatusFilter = (raw: string | null | undefined): MediaStatusToken[] =>
    normaliseStatusTokens(
        (raw ?? '')
            .split(',')
            .map((token) => token.trim().toLowerCase())
            .filter(Boolean)
    );

export const serialiseStatusFilter = (tokens: readonly string[]): string | null => {
    const normalised = normaliseStatusTokens(tokens);
    return normalised.length > 0 ? normalised.join(',') : null;
};

export const statusFilterHasFav = (raw: string | null | undefined): boolean =>
    (raw ?? '')
        .split(',')
        .some((token) => token.trim().toLowerCase() === 'fav');
